### PR TITLE
Add `Table` component to design system

### LINKED ISSUE #17: Data table component

### PR DESCRIPTION
**Pull request by rudream**
## Purpose

This PR closes https://github.com/gravitational/design-system/issues/17

Requires #109 

This PR adds the `Table` component to the design system, this table supports sorting, searching, and serverside + clientside pagination. I tried to keep the behaviour the same, however under the hood this uses `tanstack/react-table` and `tanstack/react-query` to manage table state (as opposed to our previous in-house approach).

I also added `msw` to test my changes and have a live example in storybook docs. I wasn't too sure about this because I'm not sure how useful this would be for other design components, so if it's not needed or desirable to have in the design system, let me know and I can remove it and just have the examples have directly hardcoded data.

### CODE REVIEW

File: src/components/table/table.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export { Table } from '@chakra-ui/react';
```
[the remark this answers is not in the document]
Comment (rudream) [reply src/components/table/table.ts:1]: Not relevant since this component hasn't been merged or used yet.

File: package.json [hunk 2]
```diff
@@ -39,11 +39,13 @@
...
     "@phosphor-icons/react": "^2.1.10",
     "next-themes": "^0.4.6",
-    "react-hook-form": "^7.71.1"
+    "react-hook-form": "^7.71.1",
+    "@tanstack/react-table": "^8.21.3"
   },
   "peerDependencies": {
     "@emotion/react": ">=11",
     "@fontsource/ubuntu": "^5.2.8",
+    "@tanstack/react-query": ">=5",
```
[the remark this answers is not in the document]
Comment (rudream) [reply package.json:49]: This is not relevant because only `teleport` and `cloud` will use this package, and we can expect both of them to use tables and thus require this dependency. There is also no backwards compatibility to worry about because this is being added here for the first time.

File: src/components/table/InputSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    },
+  },
+});
+
+const Form = chakra('form', {
+  base: {
+    position: 'relative',
+    display: 'flex',
+    overflow: 'hidden',
+    borderRadius: '200px',
```
Comment (ryanclark) [inline src/components/table/InputSearch.tsx]: ?

File: .storybook/preview.tsx [hunk 4]
```diff
@@ -115,8 +124,10 @@ const preview: PreviewWithStorySort = {
             'Slot Recipes',
             'Adding Chakra Components',
           ],
-          Components: ['Buttons'],
+          Components: ['Buttons', 'Forms', 'Data', 'Layout', 'Typography'],
           'Components/Buttons': ['Button', 'Close Button'],
+          'Components/Forms': ['Input', 'Field', 'Checkbox', 'Radio', 'Toggle'],
+          'Components/DataTable': ['DataTable'],
```
Comment (ryanclark) [inline .storybook/preview.tsx]: Let's make this `Components/Data Display/Data Table`

File: src/storybook/stories/components/data/Label.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { HStack } from '@chakra-ui/react';
+import type { Meta } from '@storybook/react-vite';
+
+import { Label } from '../../../../components';
+
+const meta = {
+  title: 'Components/DataTable/Label',
```
Comment (ryanclark) [inline src/storybook/stories/components/data/Label.stories.tsx]: Should label be under data table?

File: src/components/table/Cells.tsx [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    >
+      {children}
+      <SortIndicator sorted={sorted} />
+    </chakra.button>
+  );
+}
+
+function SortIndicator({ sorted }: { sorted: false | 'asc' | 'desc' }) {
+  if (sorted === 'desc') return <CaretDownIcon aria-label="sort items desc" />;
+  if (sorted === 'asc') return <CaretUpIcon aria-label="sort items asc" />;
```
Comment (ryanclark) [inline src/components/table/Cells.tsx]: I think we should remove the `aria-label` here, add support for `aria-hidden` prop and do `<SortIndicator sorted={sorted} aria-hidden />`

And then in DataTable.tsx do something like

```tsx
{headerGroup.headers.map(header => {
    const sorted = header.column.getIsSorted();
    const ariaSort =
      sorted === 'asc' ? 'ascending'
      : sorted === 'desc' ? 'descending'
      : header.column.getCanSort() ? 'none'
      : undefined;

    return (
      <Table.ColumnHeader
        key={header.id}
        colSpan={header.colSpan}
        aria-sort={ariaSort}
      >
        {header.isPlaceholder
          ? null
          : flexRender(header.column.columnDef.header, header.getContext())}
      </Table.ColumnHeader>
    );
  })}
```

**Review by ryanclark — APPROVED** [review]

File: .storybook/public/mockServiceWorker.js
Comment (bl-nero) [inline .storybook/public/mockServiceWorker.js:1]: Looks like this file has been imported from the MSW project; is there a particular reason why we need to embed it in our source code? I looked specifically for that in MSW docs, and found nothing about this procedure.

Answering your doubt about whether msw will be useful: sooner or later, it will, so it's a good call that you added it. :)

File: src/components/table/DataTable.tsx [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+  pagerPosition: PagerPosition | undefined,
+  rowCount: number
+) {
+  const hasSufficientData = rowCount > 5;
+  const showBottomPager = pagerPosition === 'bottom';
+  const showTopPager =
+    pagerPosition === 'top' || (!pagerPosition && !hasSufficientData);
+  const showBothPager =
+    pagerPosition === 'both' || (!pagerPosition && hasSufficientData);
+  return { showTopPager, showBottomPager, showBothPager };
```
Comment (bl-nero) [inline src/components/table/DataTable.tsx]: This should be `showBothPagers`

File: src/components/table/DataTable.tsx [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+              {headerGroup.headers.map(header => {
+                const sorted = header.column.getIsSorted();
+                const ariaSort =
+                  sorted === 'asc'
+                    ? 'ascending'
+                    : sorted === 'desc'
+                      ? 'descending'
+                      : header.column.getCanSort()
+                        ? 'none'
+                        : undefined;
```
Comment (bl-nero) [inline src/components/table/DataTable.tsx]: Can we extract it to a separate function (and perhaps use switch statement instead)? This is hard to read.

Also: I don't think there's any difference between `'none'` and `undefined`, at least according to MDN docs.

File: src/components/table/DataTable.tsx [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+  row,
+  isInitialLoading,
+  emptyText,
+  colSpan,
+}: {
+  rows: Row<T>[];
+  row?: DataTableRowOptions<T>;
+  isInitialLoading: boolean;
+  emptyText: string;
+  colSpan: number;
```
Comment (bl-nero) [inline src/components/table/DataTable.tsx]: It's pretty much impossible to understand the semantics of the `colSpan` attribute without digging into the code of `renderBody`; a comment would be appreciated. Or perhaps just rename the parameter to `numColumns`? You could also infer the number of colums from the first row.

File: src/components/table/DataTable.tsx [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+
+function renderBody<T>({
+  rows,
+  row,
+  isInitialLoading,
+  emptyText,
+  colSpan,
+}: {
+  rows: Row<T>[];
+  row?: DataTableRowOptions<T>;
```
Comment (bl-nero) [inline src/components/table/DataTable.tsx]: How about `rowOptions`? Otherwise, it could confuse someone who would have thought it's a special case for rendering a single row (as there's already a multi-row `rows` property).

File: src/components/table/InputSearch.tsx [hunk 11]
```diff
@@ -0,0 +1,121 @@
+import { chakra } from '@chakra-ui/react';
+import { useEffect, useState, type FormEvent, type ReactNode } from 'react';
+
+export interface InputSearchProps {
+  searchValue: string;
+  setSearchValue: (searchValue: string) => void;
+  children?: ReactNode;
+  bigInputSize?: boolean;
```
Comment (bl-nero) [inline src/components/table/InputSearch.tsx]: I'd use an `inputSize` instead, with constants of "big" and "small" ("medium"?). That's because if somone wants to use more than two sizes, it's gonna be a problem from the API side. Besides, it would be more consistent with other places in our former design system (and I guess in new one, too), where we preferred using enums for sizes.

File: src/components/table/useDataTableQuery.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+export interface DataTableServerSidePagination {
+  /** Gets the previous page. */
+  onPrev?: () => void;
+  /** Gets the next page. */
+  onNext?: () => void;
+  isLoading: boolean;
+  /** Index of the first row on the current page. */
+  from: number;
+  /** Index of the last row on the current page. */
```
Comment (bl-nero) [inline src/components/table/useDataTableQuery.ts]: Small technicality, just to make sure: is it the last row index, not the index of the next one after that, right? If so, then in case of empty page, it would mean `to` would be `from-1`, and in the corner case where we display no data at all, `to` would be actually negative. That may be OK, but something tells me it would be counterintuitive. Up to you.

**Review by bl-nero — COMMENTED** [review]
I have a couple of minor comments below and one more significant: this PR doesn't have any unit tests, and I'd love to at least see them for the client-side paging, since there's a lot of non-trivial logic and edge cases there.

File: .storybook/public/mockServiceWorker.js
Comment (rudream) [reply .storybook/public/mockServiceWorker.js:1]: MSW requires the service worker to be defined and kept in the public folder, see `Generate service worker for MSW in your public folder.` in the docs [here.](https://storybook.js.org/addons/msw-storybook-addon)

**Review by bl-nero — APPROVED** [review]
